doc/sphinx/addendum/rewrite-rules.rst

```diff
@@ -0,0 +1,144 @@
+.. _rewrite_rules:
+
+User-defined rewrite rules
+==========================
+
+.. warning::
+
+   The status of rewrite rules is highly experimental.
+
+   In particular, ill-typed rewrite rules will lead to mistyped expressions,
+   and manipulating these will most often result in anomalies.
+   Furthermore, some parts of Coq like conversion checking through bytecode or native code
+   compilation currently do not understand rewrite rules.
+
+
+The use of rewrite rules is guarded behind a flag, which can be enabled
+either by passing ``-allow-rewrite-rules`` to the
+Coq program or by turning the :flag:`Allow Rewrite Rules` flag on.
+
+.. flag:: Allow Rewrite Rules
+
+   This :term:`flag` enables the use of rewrite rules.
+   It is disabled by default and cannot be disabled once enabled.
+   The command-line flag ``-allow-rewrite-rules`` enables rewrite rules at
+   startup.
+
+   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
+      :undocumented:
+
+Declaring symbols
+-----------------
+
+Rewrite rules operate on symbols, which are their own kind of constants.
+
+.. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }
+   :name: Symbol; Symbols
+
+   Symbols are declared in the same way as axioms, by giving their type and using the keyword ``Symbol``.
+
+   .. coqtop:: all
+
+      Set Allow Rewrite Rules.
+      Symbol pplus : nat -> nat -> nat.
+      Notation "a ++ b" := (pplus a b).
+
+Declaring rules
+---------------
+
+Rewrite rules have two parts named pattern (lhs) and production (rhs).
+When a rule is applied, the term is matched against the pattern, subterms aligned with pattern holes are collected
+and substitued into the production which is returned.
+
+.. cmd:: Rewrite {| Rule | Rules } @ident := @rewrite_rule {* with @rewrite_rule }
+  :name: Rewrite; Rule; Rules
+
+   Rewrite rules are declared in blocks, which bear a name.
+
+  .. prodn::
+     rewrite_rule ::= @term ==> @term
+
+
+  .. coqtop:: all
+
+     Rewrite Rule pplus_rewrite :=
+          ?n ++ 0 ==> ?n
+     with ?n ++ S ?m ==> S (?n ++ ?m)
+     with 0 ++ ?n ==> ?n
+     with S ?n ++ ?m ==> S (?n ++ ?m).
+
+Pattern syntax
+--------------
+
+Patterns are read from a subpart of terms, with the following syntax :
+
+  .. prodn::
+     pat ::= @qualid {? @univ_annot }
```

Comment:
`pat` is too short and generic a name and can be confused with `pattern` nonterminals.  Maybe `rw_pat` instead?

Reply:
I opted for `rw_pattern`